controllers/keystoneapi_controller.go

```diff
@@ -1361,9 +1376,107 @@ func (r *KeystoneAPIReconciler) ensureFernetKeys(
 	} else {
 		// add hash to envVars
 		(*envVars)[secret.Name] = env.SetValue(hash)
-	}
 
-	// TODO: fernet key rotation
+		changedKeys := false
+
+		extraKey := fmt.Sprintf("FernetKeys%d", numberKeys)
+
+		//
+		// Fernet Key rotation
+		//
+		if secret.Annotations == nil {
+			secret.Annotations = map[string]string{}
+		}
+		rotatedAt, err := time.Parse(time.RFC3339, secret.Annotations[fernetAnnotation])
+
+		var duration int
+		if instance.Spec.FernetRotationDays == nil {
+			duration = keystone.DefaultFernetRotationDays
+		} else {
+			duration = int(*instance.Spec.FernetRotationDays)
+		}
+
+		if err != nil {
+			changedKeys = true
+		} else if rotatedAt.AddDate(0, 0, duration).Before(now) {
+			secret.Data[extraKey] = secret.Data["FernetKeys0"]
+			secret.Data["FernetKeys0"] = []byte(keystone.GenerateFernetKey())
```

Comment:
this doesn't look right. It moves key 0 to the end, and replaces it with the new key. Should replace it with key 1... etc... and FernetKeys<numberKeys-1> should be the new key